core/drivers/clk/clk.c

```diff
@@ -0,0 +1,305 @@
+
+			return res;
+		}
+	}
+
+	refcount_set(&clk->enabled_count, 1);
+
+	return TEE_SUCCESS;
+}
+
+TEE_Result clk_enable(struct clk *clk)
+{
+	uint32_t exceptions = 0;
+	TEE_Result res = TEE_ERROR_GENERIC;
+
+	exceptions =  cpu_spin_lock_xsave(&clk_lock);
+	res = clk_enable_no_lock(clk);
+	cpu_spin_unlock_xrestore(&clk_lock, exceptions);
+
+	return res;
+}
+
+void clk_disable(struct clk *clk)
+{
+	uint32_t exceptions = 0;
+
+	exceptions =  cpu_spin_lock_xsave(&clk_lock);
+	clk_disable_no_lock(clk);
+	cpu_spin_unlock_xrestore(&clk_lock, exceptions);
+}
+
+unsigned long clk_get_rate(struct clk *clk)
+{
+	return clk->rate;
+}
+
+static TEE_Result clk_set_rate_no_lock(struct clk *clk, unsigned long rate)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	unsigned long parent_rate = 0;
+
+	if (clk->parent)
+		parent_rate = clk_get_rate(clk->parent);
+
+	res = clk->ops->set_rate(clk, rate, parent_rate);
+	if (res)
+		return res;
+
+	clk_compute_rate_no_lock(clk);
+
+	return TEE_SUCCESS;
+}
+
+TEE_Result clk_set_rate(struct clk *clk, unsigned long rate)
+{
+	uint32_t exceptions = 0;
+	TEE_Result res = TEE_ERROR_GENERIC;
+
+	if (!clk->ops->set_rate)
+		return TEE_ERROR_NOT_SUPPORTED;
+
+	exceptions =  cpu_spin_lock_xsave(&clk_lock);
+
+	if (clk->flags & CLK_SET_RATE_GATE && clk_is_enabled_no_lock(clk))
+		res = TEE_ERROR_BAD_STATE;
+	else
+		res = clk_set_rate_no_lock(clk, rate);
+
+	cpu_spin_unlock_xrestore(&clk_lock, exceptions);
+
+	return res;
+}
+
+struct clk *clk_get_parent(struct clk *clk)
+{
+	return clk->parent;
+}
+
+static TEE_Result clk_get_parent_idx(struct clk *clk, struct clk *parent,
+				     size_t *pidx)
+{
+	size_t i = 0;
+
+	for (i = 0; i < clk_get_num_parents(clk); i++) {
+		if (clk_get_parent_by_index(clk, i) == parent) {
+			*pidx = i;
+			return TEE_SUCCESS;
+		}
+	}
+	EMSG("Clock %s is not a parent of clock %s", parent->name, clk->name);
+
+	return TEE_ERROR_BAD_PARAMETERS;
+}
+
+static TEE_Result clk_set_parent_no_lock(struct clk *clk, struct clk *parent,
+					 size_t pidx)
+{
+	__maybe_unused TEE_Result pres = TEE_ERROR_GENERIC;
+	TEE_Result res = TEE_ERROR_GENERIC;
+	struct clk *prev_parent = NULL;
+	bool was_enabled = false;
+	size_t prev_pidx = 0;
+
+	/* Requested parent is already the one set */
+	if (clk->parent == parent)
+		return TEE_SUCCESS;
+
+	was_enabled = clk_is_enabled_no_lock(clk);
+	/* Call is needed to decrement refcount on current parent tree */
+	if (was_enabled)
+		clk_disable_no_lock(clk);
+
+	res = clk->ops->set_parent(clk, pidx);
+	if (res)
+		goto out;
+
+	clk->parent = parent;
+
+	/* The parent changed and the rate might also have changed */
+	clk_compute_rate_no_lock(clk);
+
+out:
+	/* Call is needed to increment refcount on the new parent tree */
+	if (was_enabled) {
+		res = clk_enable_no_lock(clk);
+		if (res) {
+			pres = clk_get_parent_idx(clk, prev_parent, &prev_pidx);
+			assert(pres == TEE_SUCCESS);
+			pres = clk->ops->set_parent(clk, prev_pidx);
+			assert(pres == TEE_SUCCESS);
```

Comment:
It now seems to me that `assert(clk_enable_no_lock(clk));` at line 273 would given the same confidence level with less code. What do you think.?

Reply:
I mean... this enables a clock that was enabled not so long ago, so if a previous call to clk_enable() succeeded in the past, there is no reason why it would fail on that same clock reference now.

Sorry for this comment after I've asked you to restore the parent clock. Maybe you latest changes introducing `clk_get_parent_idx()` were not really needed.

Reply:
> It now seems to me that assert(clk_enable_no_lock(clk)); at line 273 would given the same confidence level with less code. What do you think.?
Calling a function inside assert is probably not a good idea, when compiled with NDEBUG, the call will be left out.
However, I'm not sure of that assumption for the confidence of the call to succeed. Let's say we switched the parent to another one and the clock is a PLL that needs to lock to the correct frequency. If incorrectly setup, the clock might never lock, so I guess there is still potentially problems for that. I'm not sure however if the wait for the PLL locking should be done in `set_parent()` callback or in the `enable()`.

Anyway, I'm ok that we could consider that `clk_enable_no_lock()` should succeed. However, I'll be more in favor of a `panic()` rather than an `assert()` for the same reason that said before (assert discarded when `CFG_TEE_CORE_DEBUG` is not set to y).

> Sorry for this comment after I've asked you to restore the parent clock. Maybe you latest changes introducing clk_get_parent_idx() were not really needed.

Don't worry, this function was actually previously present in clk_dt.c and has been removed thanks to the modification of clk_set_parent arguments. I also really think it make more sense for the public interface to use a `struct clk *parent` rather than a index.

Reply:
I decided to add a `if (res) panic()`after clock enabling. IMHO, this "should" not happen but let's be cautious.